### ISSUE TITLE
documentation and samples are out of date for (2.1.X/master)

### ISSUE THREAD

**Issue by navid-dada** [angular-ui/angular-google-maps#1332]
hi 
i found your api very useful and good but its documentation and samples are very small and incomplete
and this has no guide to use this library with advance features. please guide me to understand 
this library. for example using models in windows directive and so creating custom windows with marker property binding are very important but i cannot find any resource   to understand and use it.

**Comment by Balrog30** [comment]
I agree with issues in documentation...

My biggest issue so far is that the change log on the website has not been kept up to date.  I use bower to install latest minor version whenever my project builds, and things have broken with no documentation on what has changed to help me debug.  Fortunately the map component of my project is not mission-critical and it is an internal-facing app.

**Comment by iffifan** [comment]
documentation for this plugin sucks.. very bad indeed.. the whole work is ruined due to shitty documentations..

**Comment by cthrax** [comment]
@iffifan the great thing is though, since it's open source, instead of complaining about it, you can fix it! Remember that folks donate their time to this project, and while not always perfect, they are trying to get things as done as best they can with the time that they have.

Unfortunately, non-productive critiques are not actionable and really detract from the time that folks *want* to spend on the project.

**Comment by navid-dada** [comment]
you are right but documentation an samples are critical parts for attracting people to use and deploy a project. library will be rejected when annoy users ...
in this case a perfect idea and a good implementation waste because of bad documentation....

**Comment by cthrax** [comment]
@navid-dada absolutely not saying there isn't room for improvement. There most certainly is, but there just has not been enough support from the nearly 5000 unique visitors each month to keep up with development and documentation. Most of the credit, for the last couple years, goes to @nmccready for keeping up with the development tasks and only a handful of others have contributed back.

So I'm saying if you see a deficiency, please take steps to fix it, rather than getting down on the folks donating their time to produce something used by thousands of people. I guarantee that you have just as much time as the folks that are contributing (which is so precious little), so a 1000 people contributing precious little time could make for a lot of improvement.

**Comment by navid-dada** [comment]
i am agree with you but when i cannot use this library because of bad documentation how can improve it or document it for others... i think if documentation was useful and pithy others could help to improving library. now documentation of this library do not help users and confuse them Of course!!!

**Comment by nmccready** [comment]
The documentation is "bad" is that there is not enough other developers contributing to documentation and or developing the code base. Riding everything out as "free" only gets you so far. I have a kid on the way and I don't have the desire to put the amount of time into this project as I once did; especially with as little help that comes from the public. 

The documentation itself is open and can be added to at will. This code base needs less; I want "xyz" please do it for me; and more "hey I dug in and here is some changes".  

I agree that the documentation is way behind. But at the same time I (and others of the team) are burnt out from stuff exactly like this. So please move on or instead actually make a difference and make it better.

**Comment by nmccready** [comment]
I could possibly find the time to do the documentation again; but what is the reward? More complaining?

**Comment by Balrog30** [comment]
@nmccready As a project maintainer on [angular-google-chart](https://github.com/angular-google-chart/angular-google-chart), I completely understand the frustration.  The project I work on doesn't have as many users as this, but most issues I've worked to resolve have been "Help me fix this weird error caused by my code", or "I want this feature please add it for me."  And angular-google-chart does not stand as a paragon of virtue for documentation, either.

My own comment above was not intended as criticism, but as a specific pain point in documentation that I was feeling at the time (out of date change log).  @iffifan on the other hand made a very unprofessional comment, and I would just like to say that not everyone in the community is so crass.

Myself, I'd love to help out by digging through the source code and resolving issues the same way I got involved with angular-google-chart.  For me there are three reasons for which I made the conscious decision not to pursue the matter any further and walk away from this project.

1. **Language.**  I can sort of read Coffee Script as the language is basically JavaScript, so I can figure it out, but the basic structure of the language is unfamiliar so it takes extra time for me to understand what I'm reading.

2. **Complexity.**  While this project is undoubtedly smaller than the Google Maps API that it wraps and depends on, it is not something that I can gain significant understanding on in a few hours.  For every open source project I've taken in as a dependency for my own projects, I've had to look through the source for something that wasn't explained well in documentation at some point (AngularJS included).  This is the first project I've tried to do that with, and just gave up.  I couldn't find what I needed to fix the problem (I was getting Stack Overflow errors in the maps API itself, and had no idea where to start).

3. **Project Health.** Poor documentation seemed to be just a symptom of a larger problem.  I know this problem is probably just a lack of contributors, but I couldn't help but feel that helping out would be an uphill battle when I saw that the current version of the Google Maps API is 3.20, and the examples were referencing 3.17 (almost a year old).

I've removed my project's dependency on angular-google-maps, and will probably re-implement the feature I was using it for with angular-google-chart and the Maps chart type.  I feel this is very unfortunate because I would have rather overcome the issue, and then contribute my solution back, but the barrier to entry was just too high.  This is not a complaint.  I am not writing this out of anger or frustration, I am just trying to give some feedback as a potential contributor that decided to walk away from a project (a decision I do not consider permanent).  I also hope that others reading this will hopefully understand that poor project documentation is usually not a matter of incompetence on the part of the maintainers. Open Source projects are often developed as a volunteer effort, and there is a limit to how much personal time developers can donate without consequences to work or family life.  I often make design decisions on angular-google-chart based on long-term maintenance cost just because myself and the original author both have families, and not enough time to maintain a big complex project.

*You're not alone.*  ui-bootstrap had issues keeping up to the demands placed on the project owner because there were not enough maintainers to keep up sorting through new issues, never mind actually fixing bugs.  There are new contributors helping out over there, and it's made a big difference.  There appears to be a lot of progress updating ui-bootstrap to angular 1.3.x, when they were stuck on 1.2.x for ages.  Before new maintainers came on board, people were getting frustrated with the lack of progress there, too.  And not everyone expresses their frustration in a constructive manner.

In conclusion to my very long comment, I would just like to say thank you for contributing a project that scratches an itch that many developers had.  It's probably not your fault that it didn't work out for me, and I'm sorry that I couldn't dedicate the time to help out.

## TL;DR

It's not easy getting started contributing to a complex project.  Frustrated programmers should be more tactful.

**Comment by nmccready** [comment]
With regards to bullet three that is super simple fix. Its right here https://github.com/angular-ui/angular-google-maps/blob/master/src/coffee/providers/map-loader.coffee#L71 . All someone had to do was search 3.17 . It is only the default version selected and it is easily overridable. It is not even really a restriction at all.

**Comment by Balrog30** [comment]
Off-topic here, but I had tried that to see if it would resolve my stack overflow.  It did, but then it produced other errors.

**Comment by nmccready** [comment]
We could remove the version restriction and leave it as comment. I believe then it would pull in the latest by default.

**Comment by wrummler** [comment]
@nmccready Would this issue be superseded by #1384?

**Comment by wrummler** [comment]
Gonna go ahead and assume it's superseded. Closing.
